src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -222,43 +226,135 @@ def _push_content_single_part(self) -> str:
+            content_type = 'text/tab-separated-values'
+            object_key = f'manifests/{uuid4()}.tsv'
+        elif self.format == 'bdbag':
+            data = self._construct_bdbag()
+            content_type = 'application/zip'
+            object_key = f'manifests/{uuid4()}.zip'
+        else:
+            assert False
+        parameters = dict(object_key=object_key,
+                          data=data,
+                          content_type=content_type)
         return self.storage_service.put(**parameters)
 
     def _construct_tsv_content(self):
         es_search = self.es_search
 
         output = StringIO()
         writer = csv.writer(output, dialect='excel-tab')
-
         writer.writerow(list(self.manifest_entries['bundles'].keys()) +
                         list(self.manifest_entries['contents.files'].keys()))
         for hit in es_search.scan():
-            self._iterate_hit(hit, writer)
+            self._iterate_hit_tsv(hit, writer)
 
         return output.getvalue()
 
-    def _iterate_hit(self, es_search_hit, writer):
+    def _construct_bdbag(self) -> dict:
+        """Create participant.tsv and sample.tsv and return as zipped BDBag."""
+        es_search = self.es_search
+        sample_tsv = StringIO()
+        sample_writer = csv.writer(sample_tsv, dialect='excel-tab')
+        sample_writer.writerow(
+            list(self.manifest_entries['contents.specimens'].keys()) +
+            list(
+                self.manifest_entries['contents.cell_suspensions'].keys()) +
+            list(self.manifest_entries['bundles'].keys()) +
+            list(self.manifest_entries['contents.files'].keys()) +
+            ['file_url'])
+
+        participant_tsv = StringIO()
+        participant_writer = csv.writer(participant_tsv, dialect='excel-tab')
+        participant_writer.writerow(['entity:participant_id'])
+
+        for hit in es_search.scan():
+            self._iterate_hit_bdbag(hit, sample_writer, participant_writer)
+
+        return self._create_zipped_bdbag({'participant.tsv': participant_tsv, 'sample.tsv': sample_tsv})
+
+    @staticmethod
+    def _create_zipped_bdbag(data: dict):
+        """
+        Writes strings of file objects to disk, creates and returns archived (zipped) BDBag.
+
+        :parameter data: keys are the file names, values the file objects
+        :returns bdbag: archived (zipped) BDBag with values of data as payloads
+        """
+        with TemporaryDirectory() as bag_path, TemporaryDirectory() as tsv_file_dir:
+            bag = bdbag_api.make_bag(bag_path)
+            data_path = os.path.join(bag_path, 'data')
+            for fname, fobj in data.items():
+                with open(os.path.join(tsv_file_dir, fname), 'w') as f:
+                    if fname == 'participant.tsv':
+                        s = fobj.getvalue()  # remove duplicates from participant.tsv
+                        # TODO: this is rather cryptic, and could be expanded for clarity
```

Comment:
Address or remove.

Reply:
Added a comment for clarification and removed the TODO.